### ISSUE TITLE
Create docker files to have a deterministic testing environment for contributors

### ISSUE THREAD

**Issue by asdomingues** [davidhealey/waistline#247]
Would it be helpful if we had a Dockerfile to easily test or reproduce small bugs?

This wouldn't replace tests in specific devices or need for _logcats_ when reporting bugs, but would improve testing new features and possibly help newcomers in setting up the environment and running the project.

I've already [created one and it is working for browser](https://github.com/asdomingues/waistline/tree/feature/create-dockerfile/Dockerfiles), but I need yet to solve some issues at the Android one. II'm planning to use the Android version to test the [fix for nutrition names](https://github.com/davidhealey/waistline/pull/244)(because browser API doesn't allow file system access). If it is useful and fits the vision for the repo I could open a PR with the Dockerfiles.

**Comment by davidhealey** [comment]
I don't know what docker is, if it's something useful we can add it to the repo but I won't be maintaining it.

**Comment by davidhealey** [comment]
> BTW, can you find a working Android building container?

You asking me?

**Comment by yarons** [comment]
Nope, the reporter :)

**Comment by asdomingues** [comment]
> I don't know what docker is, if it's something useful we can add it to the repo but I won't be maintaining it.

Hey, I should have at least described what Docker is: my bad!

It is a tool to build containers (similar to virtual machines) by describing how they should be "lifted up" in a Dockerfile. For example: telling which operational system to use, which tools to install (like openjdk8 to use android SDK) and what to run once the container is up. It is useful in my case because I don't have Gradle, Android Studio and others installed in my machine and if I used them for other projects there could be conflicts in version, etc. But I do have docker installed and once I build and run the image I can compile the project, use its browser version via localhost and so on.
[Docker overview](https://docs.docker.com/get-started/overview/)

I was building the image with a Linux called Alpine for it to be lightweight, but that might make maintenance harder... We could swap it to Ubuntu, then maintaining should be just keeping the correct Gradle and Android SDK version at the Dockerfile, as it is used only to build the app, not in runtime or anything. That said, this is yet another thing to maintain indeed... I could get it working but if it stops at some time and I can't help fixing, it could be removed easily without breaking anything.

> We can start with @asdomingues examples, BTW, can you find a working Android building container?

I used some to get an idea of how it should be done, but didn't try to boot any as they looked too specific. I'm not on my computer right now, but I can list them as references later! :slightly_smiling_face:

**Comment by asdomingues** [comment]
It's now also working for Android. I need to do some code gardening and to describe how it was done correctly yet. But it works: I've generated an Android App and ran it successfully on my device. Although there is that unsigned app notice that @davidhealey mentioned once. Now I think it is possible to compile different versions of the app according to the commits and test db migration.

I've used this [repo](https://github.com/alvr/alpine-android) to generate the base image and this [Dockerfile](https://github.com/chickenzord/docker-alpine-gradle/blob/master/dockerfiles/2.10/Dockerfile) as a reference to include Gradle. [This article](https://swaminathanvetri.in/2016/12/24/building-cordova-apps-using-docker-images/) on creating docker images for cordova apps also helped.
